synapseclient/table.py

```diff
@@ -780,6 +781,119 @@ def __init__(self, name=None, columns=None, parent=None, definingSQL=None, prope
         )
 
 
+class Dataset(SchemaBase):
+    """
+    A Dataset is an :py:class:`synapseclient.entity.Entity` that defines a
+    flat list of entities as a tableview (a.k.a. a "dataset").
+
+    :param name:            The name for the Dataset object
+    :param description:     User readable description of the schema
+    :param columns:         A list of :py:class:`Column` objects or their IDs
+    :param parent:          The Synapse Project to which this Dataset belongs
+    :param properties:      A map of Synapse properties
+    :param annotations:     A map of user defined annotations
+    :param dataset_items:    A list of items characterized by entityId and versionNumber
+    :param local_state:     Internal use only
+
+    Example::
+
+        from synapseclient import Dataset
+        # Create a Dataset with pre-defined DatasetItems
+        dataset_items = [
+            {'entityId': "syn000", 'versionNumber: 1},
+            {...},
+        ]
+        dataset = syn.store(Dataset(
+            name="My Dataset",
+            parent=project,
+            dataset_items=dataset_items))
+
+        # Add/remove specific Synapse IDs to/from the Dataset
+        dataset.add_item({'entityId': "syn111", 'versionNumber': 1})
+        dataset.remove_item("syn000")
+
+        # Add a list of Synapse IDs to the Dataset
+        new_items = [
+            {'entityId': "syn222", 'versionNumber': 2},
+            {'entityId': "syn333", 'versionNumber': 1}
+        ]
+        dataset.add_items(new_items)
+
+    To create a snapshot version of the Dataset, use
+    :py:classmethod:`synapseclient.client.create_snapshot_version`.
+
+    Example::
+        syn = synapseclient.login()
+        syn.create_snapshot_version(
+            dataset.id,
+            label="v1.0",
+            comment="This is version 1")
+    """
+    _synapse_entity_type: str = "org.sagebionetworks.repo.model.table.Dataset"
+    _property_keys: List[str] = SchemaBase._property_keys + ['dataset_items']
+
+    def __init__(self, name=None, columns=None, parent=None, properties=None,
+                 annotations=None, local_state=None, dataset_items=None, **kwargs):
+        self.properties.setdefault('dataset_items', [])
+        super(Dataset, self).__init__(
+            name=name, columns=columns, properties=properties,
+            annotations=annotations, local_state=local_state, parent=parent,
+            **kwargs
+        )
+        if dataset_items:
+            self.add_items(dataset_items)
+
+    def __len__(self):
```

Comment:
Love this addition.

Reply:
Oh yay!  I wasn't sure if overwriting the dunder method was okay but in my mind, len() should return the number of items in the dataset, rather than the number of entity properties.